ros/src/waypoint_updater/waypoint_updater.py

```diff
@@ -94,17 +95,51 @@ def waypoints_cb(self, waypoints):
             self.waypoints_2d = [[waypoint.pose.pose.position.x, waypoint.pose.pose.position.y] for waypoint in
                                  waypoints.waypoints]
             self.waypoints_tree = KDTree(self.waypoints_2d)
-            self.waypoint_distance = self.point_distance(self.waypoints_2d[0], self.waypoints_2d[1])
 
-    def publish_waypoints(self, closest_waypoint_idx):
+
+    def publish_waypoints(self):
+        if self.waypoints_2d:
+            final_lane = self.generate_lane()
+            self.final_waypoints_pub.publish(final_lane)
+
+    def generate_lane(self):
         lane = Lane()
         lane.header = self.base_waypoints.header
-        lane.waypoints = self.base_waypoints.waypoints[closest_waypoint_idx: closest_waypoint_idx + LOOKAHEAD_WPS]
-        self.final_waypoints_pub.publish(lane)
+        closest_waypoint_idx = self.get_closest_waypoint_idx()
+        farthest_idx = closest_waypoint_idx + LOOKAHEAD_WPS
+        the_waypoints = self.base_waypoints.waypoints[closest_waypoint_idx: farthest_idx]
+
+        if self.stopline_wp_idx == -1 or (self.stopline_wp_idx >= farthest_idx):
+            lane.waypoints = the_waypoints
+        else:
+            lane.waypoints = self.decelerate_waypoints(the_waypoints, closest_waypoint_idx)
+
+        return lane
+        
+    def decelerate_waypoints(self, waypoints, closest_idx):
+        temp = []
```

Comment:
we may want a better name. Seems like the result array.

Reply:
I think this variable name was copied from the classroom video, seems to be a minor issue.

Reply:
I'll update after pulling in. This will be public with our name on it...